### ISSUE TITLE
Multi user environments

### ISSUE THREAD

**Issue by lociii** [jazzband/django-simple-menu#27]
There is a bug in multi user environments:

The visibility of an item does not get reset between requests of different users. Therefore a user with a lot of privileges will not see all of it's menu items when a user with less privileges already populated the visible flag of a menu item.
It's not possible to restore the default value of visible since it is overwritten by the check callback.

**Comment by georgema1982** [comment]
I wonder why this hasn't been merged into master. It's not only affecting multi-user but also single user as well. I'm experiencing a bug under single user senario that's related to it. Basically, a parent menu item has no check but its submenu items have checks regarding permissions. I set MENU_HIDE_EMPTY because I want to hide the parent menu if none of its children are visible. Initially, when the user is under logged-in state, the menu display properly. But once he logs out, the parent menu's visible is set to False because none of its children are visible, which is correct. Now I login again. Due to the fact that parent menu's visible isn't reset, it won't show up any more.

**Comment by borgstrom** [comment]
I'm traveling the next couple of days and won't have a chance to look into this until I'm back.  TBH, the reason this wasn't merged is that the original reporter closed the issue before I could get to it. 

Do you think you could try to write a test that shows the issue you're experiencing?  It will make merging a fix much easier. 

P.S.  I haven't forgotten about your other issue either.  I'm all for you being able to not have to repeat yourself in checks and will respond to it when I'm back as well. 

Thanks!

**Comment by georgema1982** [comment]
Sorry, not to be meant to repeat this issue. But after reading the source codes further, I wonder if it's thread safe at all. Thread safety might be the root cause.

From what I see, MenuItems are loaded only once and is set on the Menu class level. Everytime Menu's process function is called, the same MenuItem instances are processed. Doesn't look thread safe to me.

**Comment by georgema1982** [comment]
Just read your source codes of this commit. I wonder if it would be better that everytime a menu is processed, new menuitem instances are created based on the loaded menuitem instances, and only the new menuitem instances are processed. That is, the menuitems defined in menus.py files only serve as a template and the processing only happens on the instances cloned from this template. My concern with this lock solution is that it won't perform well under multi-user environment.

**Comment by borgstrom** [comment]
Why do you think reading all of the menus.py files on every request will be faster than simply acquiring and releasing a lock?

Moving to per-request generation of the items will require significant work to maintain backwards compatibility.  I would need more convincing that it's the better path forward.

**Comment by georgema1982** [comment]
I don't think reading menus.py files on each request necessary. Load it only once. But on each request, clone all the menuitem instances from the loaded ones in the memory and process the cloned ones. Lock vs new instance are definitely design preferences. But I think the new instance looks more straight forward because each request should have their own menuitems instances to process just like each request should have their own model instances to process in the views.

**Comment by georgema1982** [comment]
Another strong reason why I prefer cloned instances is because the lock doesn't solve lociii or my issues. Why? Because there are some attributes of MenuItem are changed during the process. The notable example is visible and selected attribute. After process, their original value might be lost. Next time you process it, you start with a tainted state of the menuitem. Lock only guarantees that multiple threads won't read and update these two attributes at the same time, but it can't prevent the tainted value.

I think the easiest fix is actually change the code below:
        for item in c.items[name]:

to:
        for item in c.items[name][:]:

It's thread safe and because it's a clone of the loaded value, it guarantee that value change won't affect the next process.
